### PR TITLE
Add critical flag

### PR DESCRIPTION
**Pull request by cabo**

### CODE REVIEW

**Review by justingrant — COMMENTED** [review]
I added a few suggestions.

File: draft-ietf-sedate-datetime-extended.md [hunk 1]
```diff
@@ -357,6 +357,57 @@ not specifically configured to take part in such an experiment.
...
+
+    2022-07-08T00:14:07Z
+
+Similar with:
+
+    2022-07-08T00:14:07Z[knort=blargel]
+
+However,
+
+    2022-07-08T00:14:07Z[!Europe/Paris]
```
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:397]: Do you have a proposal for a critical-flag that is forward-compatible with JAVAZDT?

File: draft-ietf-sedate-datetime-extended.md [hunk 2]
```diff
@@ -357,6 +357,57 @@ not specifically configured to take part in such an experiment.
...
+
+A suffix tag may also indicate that it is *critical*: The recipient is
+advised that it MUST not act on the Internet Date/Time Format string
+unless it can process the suffix tag as specified.  A critical suffix
+tag is indicated by following its opening bracket with an exclamation
+mark (see `critical-flag` in {{abnf}}).
+
+A string such as:
+
+    2022-07-08T00:14:07Z[Europe/Paris]
```
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:381]: This seems to be a proposal to interpret timezone hints in a different way than I was thinking we were.  So we first would need to discuss that new interpretation.

Note that RFC 3339 is quite specific on this:

```
4.3. Unknown Local Offset Convention

   If the time in UTC is known, but the offset to local time is unknown,
   this can be represented with an offset of "-00:00".  This differs
   semantically from an offset of "Z" or "+00:00", which imply that UTC
   is the preferred reference point for the specified time.  RFC2822
   [IMAIL-UPDATE] describes a similar convention for email.
```

File: draft-ietf-sedate-datetime-extended.md [hunk 3]
```diff
@@ -357,6 +357,57 @@ not specifically configured to take part in such an experiment.
...
+
+is internally inconsistent, as Europe/Paris does not use a time zone
+offset of 0.
+The time zone hint given is elective, though, so the recipient is not
+required to act on the inconsistency; it can treat the Internet
+Date/Time Format string as if it were:
+
+    2022-07-08T00:14:07Z
```
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:389]: I would like to keep the Z here exactly because it is also an illustration of the semantics of Z as a time zone offset.

File: draft-ietf-sedate-datetime-extended.md [hunk 4]
```diff
@@ -357,6 +357,57 @@ not specifically configured to take part in such an experiment.
...
+
+Similar with:
+
+    2022-07-08T00:14:07Z[knort=blargel]
```
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:393]: I would like to keep the Z here exactly because it is also an illustration of the semantics of Z as a time zone offset.

File: draft-ietf-sedate-datetime-extended.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:397]: I would like to keep the Z here exactly because it is also an illustration of the semantics of Z as a time zone offset.

File: draft-ietf-sedate-datetime-extended.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:381]: It would be ideal if implementers had adopted RFC 3339's use of `-00:00` for all offset-unknown timestamps. But in practice, I've been unable to find any major computing platform with a timestamp data type that uses that format. Sadly, the only cases I've found use `Z`, not `-00:00` when serializing offset-less timestamps.

How have other standards dealt with cases where implementers have consistently ignored the standard in the same consistent way?  Do the standards sometimes update with a new standard that matches what implementers have consistently decided to do?

To help with this discussion, below are all the platforms I found that:
* a) have an offset-less timestamp data type AND ... 
* b) ... that type has a built-in serialization into ISO 8601 and/or RFC 3339 formats AND ...
* c) ... that serialization method does not require a time zone or offset as an input.

All of them use `Z` and none use `-00:00`, including those which explicitly claim to be supporting RFC 3339!

## Java

From the [docs](https://docs.oracle.com/javase/10/docs/api/java/time/format/DateTimeFormatter.html#ISO_INSTANT):
> The ISO instant formatter that formats or parses an instant in UTC, such as '2011-12-03T10:15:30Z'.

Java serializes with `Z` and throws an exception if you try to parse a timestamp with `-00:00`:
```java
import java.time.Instant;
class Main {
  public static void main(String[] args) {
    Instant instantOK = java.time.Instant.parse("2020-01-01T00:00:00Z");
    System.out.println(instantOK.toString());
    // 2020-01-01T00:00:00Z

    Instant instantFail = java.time.Instant.parse("2020-01-01T00:00:00-00:00");
    System.out.println(instantFail.toString());
    // Exception in thread "main"
    // java.time.format.DateTimeParseException: 
    // Text '2020-01-01T00:00-00:00' could not be
    //   parsed at index 16
  }
}
```

## Kotlin

Unsurprisingly, Kotlin (Android's variant of Java) behaves similarly to Java, but at least it will parse `-00:00` without an exception. Its output still uses `Z` though.

```kotlin
import java.time.Instant;
fun main() {
    var instant1: Instant = java.time.Instant.parse("2020-01-01T00:00:00Z");
    System.out.println(instant1.toString());
    // 2020-01-01T00:00:00Z

    var instant2: Instant = java.time.Instant.parse("2020-01-01T00:00:00-00:00");
    System.out.println(instant2.toString());
    // 2020-01-01T00:00:00Z
}
```

## Swift 

From the [docs](https://developer.apple.com/documentation/foundation/iso8601dateformatter/options/1643217-withinternetdatetime):

> The format used for internet date times, according to the [RFC 3339](https://www.ietf.org/rfc/rfc3339) standard.

Serializes with `Z`. Parsing not tested.

```swift
import Foundation
let formatter = ISO8601DateFormatter()
formatter.formatOptions = [.withInternetDateTime] // use "RFC 3339" format
print(formatter.string(from: Date()))
// 2022-07-09T18:38:20Z
```

## (Pre-Temporal) JavaScript

```js
new Date().toISOString()
// '2022-07-09T22:34:10.264Z'
```

## Go

```go
package main
import ( 
  "fmt"
  "time"
)

func main() {
  fmt.Println(time.Now().Format(time.RFC3339))
  // 2022-07-09T22:42:43Z
}
```

## .NET (using NodaTime library)

The .NET standard library doesn't have an exact-time type, but [NodaTime](https://nodatime.org/) is a very commonly-used .NET library that does have one: [NodaTime.Instant](https://nodatime.org/3.1.x/api/NodaTime.Instant.html). 

```C#
using System;
using NodaTime;			
public class Program
{
	public static void Main()
	{
		Instant now = SystemClock.Instance.GetCurrentInstant();
		Console.WriteLine(now.ToString());
                 // 2022-07-09T23:07:02Z
	}
}
```

## MongoDB

This DBMS has an exact-time type called Date. Its default string serialization uses `Z` ([docs](https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/#mongodb-bsontype-Date:~:text=%5B%22hello%22%2C10%5D-,%22dateField%22%3A,-%7B%22%24date%22%3A%7B%22%24numberLong%22%3A%221565546054692)) but, weirdly, only for dates after 1970.  Earlier dates use integers.  

Here's some query output from the MongoDB console when querying a document with a `Date`.

```js
[
  {
    "_id": ObjectId("5a934e000102030405000000"),
    "d": ISODate("2021-04-30T01:36:17Z")
  }
]
```

[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:381]: Thank you for this survey.
So we know that many date/time implementations are not expressive enough to distinguish between a (UTC-referenced) timestamp with a UTC time zone offset and a (UTC-referenced) timestamps with an unknown time zone offset.
RFC 3339 does make that distinction, though.
Since we are supposed to be an extension of RFC 3339, we will have to handle it.

2021-04-30T01:36:17Z[Europe/Paris] has a +00:00 time zone offset and a conflicting time zone hint.
A conflict handler in an application could of course decide that the evidence goes in favor of the time zone hint.
The regions suffering would be of small economic impact (Reykjavik, Timbuktu, etc.), but that is not a great way to handle this.
So I don't think we have any good way to make use of the information you gathered.

File: draft-ietf-sedate-datetime-extended.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:389]: I instead added a non-zero time zone offset as well.

File: draft-ietf-sedate-datetime-extended.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:381]: I agree that there's no perfect solution here. But I think there's a "least bad" solution: codify existing practice among the (AFAICT) vast majority of modern computing platforms which use `Z` instead of `-00:00`. 

Specifically:
* `2022-07-08T00:14:07Z[!Europe/Paris]` MUST NOT cause an error
* `2022-07-08T00:14:07+00:00[!Europe/Paris]` MUST cause an error
* `2022-07-08T00:14:07Z[Europe/Paris]` SHOULD NOT cause an error in applications that do conflict detection
* `2022-07-08T00:14:07+00:00[Europe/Paris]` SHOULD cause an error in applications that do conflict detection

The SHOULDs above acknowledge the case that when interoperating with existing legacy software that uses this format, it may not be possible to change the other party's behavior. But the critical flag is brand-new, so there's no legacy and MUST is appropriate.

> A conflict handler in an application could of course decide that the evidence goes in favor of the time zone hint.

If we can't get consensus on the codify-existing-practice interpretation of `Z`, then at a minimum I think we should include a statement like yours above in the standard, so that application developers will have some guidance on how to handle the disconnect between the software platforms they use and the new standard.  Maybe something like this?

> **Offset vs. Time Zone Conflicts**
> 
> Time zone offsets in a timestamp string can conflict with its time zone suffix.
> For example, this could happen if the definition of a time zone is changed (for example, to abolish Daylight Saving Time) after a timestamp string is stored that represents a far-future event.
> If time zone and offset are in conflict, even if the critical flag is not used for the time zone suffix, an application MAY reject the timestamp or resolve the conflict via user and/or programmer input.
>
> ** Use of `Z` Offset with Time Zone Suffix **
>
> RFC 3339 (as well as this standard that extends it) provides the offset `-00:00` to indicate that the actual local time and local offset are unknown.
> However, many implementations use `Z` for this purpose, often to maximize compatibility with ISO 8601 which disallows `-00:00`.
> Therefore, implementations MAY choose to interpret `Z` the same as `-00:00`: exempting `Z` offsets from conflict detection and instead using the time zone suffix to determine the local offset and local time.
> Because of this common behavior, implementations SHOULD use `+00:00` instead of `Z` when recording a timestamp when the local UTC offset is known to be zero and SHOULD NOT use `+00:00` when the local UTC offset is unknown.

File: draft-ietf-sedate-datetime-extended.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:412]: Nope, the Java parser is very picky (e.g. won't even accept `-00:00`) so any change to the string format won't be accepted by JAVAZDT.  (Although of course a later version of Java's standard library could support it.)

I'm mostly wondering if it's worth adding a note warning clients that the critical flag is new syntax that will be incompatible with existing users of the time zone suffix like JAVAZDT.  If this kind of warning isn't usually done in IETF specs, then IMO it's OK to omit.

File: draft-ietf-sedate-datetime-extended.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:393]: I don't have a copy of ISO 8601, so I can't check the status of -00:00 in that document.  I'm a bit surprised that RFC 3339 has a major feature that is disallowed by ISO 8601.

[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:393]: Here's what Wikipedia says, with citations. 

> An offset of zero, in addition to having the special representation "Z", can also be stated numerically as "+00:00", "+0000", or "+00". However, it is not permitted to state it numerically with a negative sign, as "−00:00", "−0000", or "−00". The section dictating sign usage[[30]](https://en.wikipedia.org/wiki/ISO_8601#cite_note-30) states that a plus sign must be used for a positive or zero value, and a minus sign for a negative value. Contrary to this rule, RFC 3339, which is otherwise a profile of ISO 8601, permits the use of "-00", with the same denotation as "+00" but a differing connotation.[[31]](https://en.wikipedia.org/wiki/ISO_8601#cite_note-31)[[32]](https://en.wikipedia.org/wiki/ISO_8601#cite_note-32)

The ISO spec is a little less definitive, but I assume that Java is interpreting it like Wikipedia does:

> [±] represents a plus sign [+] if in combination with the following element a positive value or zero needs to be represented (in this case, unless explicitly stated otherwise, the plus sign shall not be omitted), or a minus sign [−] if in combination with the following element a negative value needs to be represented.

[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:393]: OK, I still don't have a copy of ISO 8601, but so many references seem to indicate that there indeed was a version of ISO 8601 that had this restriction.  It would be interesting to verify this against ISO 8601:2019.

Still, our reference is RFC 3339, and that is very explicit about what -00:00 means.

[the remark this answers is not in the document]
Comment (cabo) [reply draft-ietf-sedate-datetime-extended.md:393]: (To put it very bluntly, in case my position on this isn't clear: the open standard RFC 3339 trumps the club standard ISO 8601 for all intents and purposes, and, more importantly, it happens to be the basis of our charter.)

[the remark this answers is not in the document]
Comment (gibson042) [reply draft-ietf-sedate-datetime-extended.md:393]: ISO 8601-1:2019 does indeed have the text about "±" meaning a required plus sign for nonnegative values and a required minus sign for negative values, and uses it in the definition of the time shift components "often used in the representation of local standard time against UTC" in section 4.3.13 (which also includes guidance disallowing -00:00: «_The time shift shall be expressed as positive (i.e. with the leading plus sign ["+"]) if it is ahead of or equal to UTC_»).

File: draft-ietf-sedate-datetime-extended.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryzokuken) [reply draft-ietf-sedate-datetime-extended.md:412]: I agree with @cabo here. This is new syntax, but the old syntax is also supported: this just adds an additional capability that might be easily added in a future version of JAVAZDT without breaking anything.

File: draft-ietf-sedate-datetime-extended.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:393]: > IIUC `2022-07-08T00:14:07` has the same connotation in ISO-land as `2022-07-08T00:14:07-00:00` in IETF-land.

@ryzokuken - Did you mean `2022-07-08T00:14:07Z` in the first string?

File: draft-ietf-sedate-datetime-extended.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:412]: > What about the [u-ca=...] feature in JAVAZDT?

Current Java won't accept any other suffixes either. But 99%+ usage in Temporal won't emit that calendar suffix either, because non-Gregorian calendars are a rarely-used feature outside of localized formatting.

Regardless, if we don't think a note is needed in this case then I'm OK with that. Resolving.

[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:412]: Oh, looks like I can't resolve conversations I started. :-) Anyway, this one seems OK & should be resolved.

File: draft-ietf-sedate-datetime-extended.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (justingrant) [reply draft-ietf-sedate-datetime-extended.md:393]: Oh!  AFAIK, those two strings actually mean different things:
* `2022-07-08T00:14:07-00:00` means "The instant is `2022-07-08T00:14:07Z` but I don't know what the local offset is, so neither the local time nor the offset are meaningful and they should not be used."
* `2022-07-08T00:14:07` means "I know that the local date is `2022-07-08` and the time is `00:14:07`, but I have no idea what the instant is that corresponds to that local date and time."

In other words, the former is a timestamp (an exact point in time), but I don't know where it was measured. The latter is *not* a timestamp... it's just a recording of the local date and time in some unknown place.  They are not interchangeable. 

@cabo does that match your understanding?
